mongo/mongo.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2021 Flamego. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package mongo
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/flamego/session"
+)
+
+var _ session.Store = (*mongoStore)(nil)
+
+// mongoStore is a MongoDB implementation of the session store.
+type mongoStore struct {
+	nowFunc    func() time.Time // The function to return the current time
+	lifetime   time.Duration    // The duration to have no access to a session before being recycled
+	db         *mongo.Database  // The database connection
+	collection string           // The database collection for storing session data
+	encoder    session.Encoder  // The encoder to encode the session data before saving
+	decoder    session.Decoder  // The decoder to decode binary to session data after reading
+}
+
+// newMongoStore returns a new MongoDB session store based on given configuration.
+func newMongoStore(cfg Config) *mongoStore {
+	return &mongoStore{
+		nowFunc:    cfg.nowFunc,
+		lifetime:   cfg.Lifetime,
+		db:         cfg.db,
+		collection: cfg.Collection,
+		encoder:    cfg.Encoder,
+		decoder:    cfg.Decoder,
+	}
+}
+
+func (s mongoStore) Exist(ctx context.Context, sid string) bool {
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Err()
+	return err != mongo.ErrNoDocuments
+}
+
+func (s mongoStore) Read(ctx context.Context, sid string) (session.Session, error) {
+	var result bson.M
+	err := s.db.Collection(s.collection).FindOne(ctx, bson.M{"key": sid}).Decode(&result)
+	if err == nil {
+		binary, ok := result["data"].(primitive.Binary)
+		if !ok {
+			return nil, errors.New("assert `data` key")
+		}
+
+		expiredAt, ok := result["expired_at"].(primitive.DateTime)
```

Comment:
```suggestion
		expiredAt, ok := result["expiredAt"].(primitive.DateTime)
```

Is MongoDB's naming convention is to use snake case? If so, please ignore this suggestion.

Reply:
It looks like in the official example they're using snake case.
https://sourcegraph.com/github.com/mongodb/docs-golang@master/-/blob/source/includes/usage-examples/code-snippets/updateOne.go?L40

Reply:
Cool, then let's stick with snake case.